src/agent/debuggable-module/src/block.rs

```diff
@@ -0,0 +1,286 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use anyhow::Result;
+use iced_x86::Decoder;
+use std::collections::{BTreeMap, BTreeSet};
+
+use crate::debuginfo::DebugInfo;
+use crate::{Module, Offset};
+
+pub fn sweep_module<'data>(module: &dyn Module<'data>, debuginfo: &DebugInfo) -> Result<Blocks> {
+    let mut blocks = Blocks::default();
+
+    for function in debuginfo.functions() {
+        let function_blocks = sweep_region(module, debuginfo, function.offset, function.size)?;
+        blocks.map.extend(&function_blocks.map);
+    }
+
+    Ok(blocks)
+}
+
+pub fn sweep_region<'data>(
+    module: &dyn Module<'data>,
+    debuginfo: &DebugInfo,
+    offset: Offset,
+    size: u64,
+) -> Result<Blocks> {
+    use iced_x86::Code;
+    use iced_x86::FlowControl::*;
+
+    let region = offset.region(size);
+
+    let data = module.read(offset, size)?;
+    let mut decoder = Decoder::new(64, data, 0);
+
+    let mut visited = BTreeSet::new();
+
+    let mut pending = Vec::new();
+
+    // Schedule the function entrypoint.
+    pending.push(offset.0);
+
+    // Schedule any extra jump labels in the target region.
+    for label in debuginfo.labels() {
+        // Don't duplicate function entrypoint.
+        if label == offset {
+            continue;
+        }
+
+        // Don't visit labels outside of the function region.
+        if !region.contains(&label.0) {
+            continue;
+        }
+
+        pending.push(label.0);
+    }
+
+    while let Some(entry) = pending.pop() {
+        if !region.contains(&entry) {
+            continue;
+        }
+
+        if visited.contains(&entry) {
+            continue;
+        }
+
+        visited.insert(entry);
+
+        // Reset decoder for `entry`.
+        let position = (entry - offset.0) as usize;
+        decoder.set_position(position)?;
+        decoder.set_ip(entry);
+
+        // Decode instructions (starting from `entry`) until we reach a block
+        // terminator or run out of valid data.
+        while decoder.can_decode() {
```

Comment:
We identify basic blocks through a two-pass sweep of the function. I could've abstracted each loop and expressed the passes with callbacks, but I think it'd be a lot harder to read.